RFC: Operators

gatesn opened the pull request:
See also https://github.com/vortex-data/vortex/issues/4492

gatesn committed:
Add operators RFC

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Add operators RFC

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Add operators RFC

Signed-off-by: Nicholas Gates <nick@nickgates.com>

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+# Array Operators
+
+This RFC proposes to turn Vortex arrays into logical query plans to support multiple modes of execution. Compute
+functions become arrays, and expressions disappear in favor of serialized arrays with placeholder values.
```
"Expressions disappear in favor of serialized arrays with placeholder values" sounds weird. I feel like what you want to say is what we currently call Expressions are really just "unbound Arrays"? or like, Arrays with unbound children?

Or perhaps this all comes back to "Arrays and Expressions are all just nodes in a Logical Plan"?

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+responsibility) over how much or little compute to actually perform.
+
+For example, `filter(DictArray, Mask)` could fully evaluate the filter, or it could push down the filter over the
+dictionary codes with an unknown cost.
```
Worth noting that the problem here is not just unpredictability... looking at this as a Logical Plan, it tends to make  optimization opportunities local and intermingled with the actual execution, since every layer is somewhat opaque.

This RFC lets us have a separate planning/optimization step

gatesn replied:
The next motivation point no?

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+## Arrays as a Logical Plan
+
+We have long described the scope of Vortex as performing “linear compute”. In other words, any compute that doesn’t
+require shuffling data. In many ways this RFC builds on the idea that really Vortex is fundamentally a subset of a query
```
I'd frame it as "Vortex implements a subset of functionality usually performed by the query engine. Specifically, we implement a highly optimized physical Scan operation, that supports pushdown of filters, projections, and scalar expressions."

It's worth being very precise here with the relational algebra / common query engine terminology

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+require shuffling data. In many ways this RFC builds on the idea that really Vortex is fundamentally a subset of a query
+engine and leans into this with somewhat familiar terminology.
+
+Vortex arrays will become nodes in a logical plan where the execution of such a plan returns a canonical array. Recall
```
Must the execution return a canonical array? How does this work for e.g., duckdb export, where we want an ALP or Dictionary vector?

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+codecs and deconstructs the array as required.
+
+For example, DuckDB supports dictionary-encoded vectors. So we would try to downcast the root node as a DictArray, and
+if successful, canonicalize the codes and values children separately.
```
This example is helpful, but it feels like we should carry it further. Since on its face, the prior section seems slightly in conflict... I think you're just missing a sentence to the effect of "This effectively splits the original execution plan into multiple execution plans, to meet the prior invariant that each execution returns a canonical array"

But then it raises the question... if they're now disjoint, how does that square with Common Subtree Elimination below?

lwwmanning commented on rfcs/0001-operators.md:
````diff
@@ -0,0 +1,369 @@
+        rhs: Constant{value = 1}
+```
+
+**Example 3: Reduce Parent - Pull Down**
````
these examples are great

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+
+    // Inspection APIs for push-down / pull-up optimization
+    fn children(&self) -> &[ArrayRef];
+    fn with_children(self: Arc<dyn Self>) -> VortexResult<ArrayRef>;
```
what does with_children do...?

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+
+Early experiments have shown up to 4x improvements in performance.
+
+Pipelines operator over vectors of 1024 elements at a time, until the final vector which may be shorter. They are not
```
typo: s/operator/operate/

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+
+During the bind phase, arrays buffers are moved over from CPU to GPU memory. Note that these buffers contain fully
+compressed data since they live in the leaves of the compressed array tree. In the future, we should be able to hint to
+the vortex-scan logic that buffers should be loaded directly from files/network into the GPU, avoiding the compressed
```
🔥

lwwmanning commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+
+## Open Questions
+
+- Is the canonical return value to performance-restrictive
```
typo: s/to/too/

blaginin commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,369 @@
+Many query engines support partially compressed data (e.g. DuckDB, Velox) rather than fully decompressed Arrow (e.g.
+DataFusion, Polars).
+
+For these engines, we propose that the export logic inspects the root node of the array for compatible compression
```
Can you please elaborate on this a bit more? 

In spiral, if there are several fragment files, we would always combine them using vortex `zip` function. Say, it would become an operator. Then we'll always need to decompress data on (almost) every single read in vortex. Then we won't be able to use, for example, FSST in spiral at all?

gatesn committed:
Merge branch 'develop' into ngates/operator-rfc

gatesn committed:
Update RFC

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Merge branch 'develop' into ngates/operator-rfc

gatesn committed:
Vector Operators

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Vector Operators

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Vector Operators

Signed-off-by: Nicholas Gates <nick@nickgates.com>

gatesn committed:
Vector Operators

Signed-off-by: Nicholas Gates <nick@nickgates.com>

connortsui20 committed:
Some operator RFC revisions (#5013)

Some rewording and additions that I think might be helpful for people
encountering this for the first time

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+  this tree are represented by the placeholder expression called `scope` (often denoted as `$`).
+* Expressions can be converted to arrays by first converting their children to arrays, then calling the `build`
+  function of the array encoding indicated by the `id` of the expression, along with the serialized `metadata`. This
+  process results in a bound array now with known `len` and `dtype`.
```
By "bound" we mean that the expression now has data bound to it, right? I think that might be worth spelling out

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+* All evaluation logic (for executing the expression), and analysis logic (for producing pruning predicates) will be
+  moved to the `Array` trait.
+
+Expressions simply provide a way of describing an unbound query over a root scope array and make up the user-facing
```
I am not sure I understand what a "scope" is super well, above it says that scopes are at the leaves of the expression tree, but this says it is at the root?

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+arrays _are_ async, and so we can remove `LayoutReader` entirely in favor of regular Vortex arrays.
+
+This change further allows us to simplify scanning logic and avoid the current expression partitioning. Instead of
+splitting expressions and pushing the partitions down over the fields of a struct layout, we simply apply the expression
+to a row-range of a layout to produce an array: `Layout::apply(Expr, RowRange) -> ArrayRef`. A `FlatLayout` would be
```
This probably stems from my lack of understanding about how expressions and layouts currently work, but I thought that in this new model we get the expression **from** our layouts, not apply a layout to the expression. Basically, where is the expression coming from if not from the layout?

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+This change further allows us to simplify scanning logic and avoid the current expression partitioning. Instead of
+splitting expressions and pushing the partitions down over the fields of a struct layout, we simply apply the expression
+to a row-range of a layout to produce an array: `Layout::apply(Expr, RowRange) -> ArrayRef`. A `FlatLayout` would be
+modelled as a node in the array tree allowing us to perform partial optimizations over the tree, such as pushing
+`get_item` expressions through a `StructLayout`.
```
Not sure I understand why `FlatLayout` being a node in the array tree is the specific reason that allows us to push `get_item` through a `StructLayout`, isn't `FlatLayout` supposed to be the leaf nodes before we are able to "resolve" it to something more specific?

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+
+## Optimizations
+
+### Common Subtree Elimination
```
I think this means that we do not have to cache _any_ computed data / never do duplicate work, is that true or are there exceptions?

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+
+## Array API
+
+This change essentially collapses the API for an array down to just an async “into_vector” function. Almost everything
```
Should `into_vector` be added to `Array` above?

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+This change essentially collapses the API for an array down to just an async “into_vector” function. Almost everything
+can be expressed via this function, and it means we can strip down the surface area required for array implementations.
+
+For example, functions relating to validity can be removed and implemented using an IsNull operator pushed down over the
+array tree and evaluated into a boolean vector. Even scalar_at could be implemented with a singleton selection array,
+evaluated to a vector, and then converted to a scalar.
```
I think that it is additionally worth considering refactoring our scalars to be a lot closer to vectors, rather than what we currently have which is opaque views into (physical) type-erased `Array`s. If we say that scalars become the elements of our vectors (in which we can additionally have references into our vectors), I bet we would have _massive_ performance gains from the reduction in (imo) wasteful allocations we do to keep everything opaque.

connortsui20 commented on rfcs/0001-operators.md:
```diff
@@ -0,0 +1,495 @@
+
+During construction, the pipeline goes through common sub-tree elimination, followed by a vector allocation stage. This
+stage computes the minimal number of vectors required for execution based on the topological sort of the pipeline DAG (
+essentially the same problem as register allocation in a compiler). By minimizing the number of intermediate vectors, we
+can maximize our use of CPU caches. Where possible, pipeline operators will write directly into the output vector.
```
Are we planning on implementing the reuse of memory (after we use a vector we can reuse its memory for a future vector, this is what `collect` does and is probably more complicated to implement?).

connortsui20 reviewed (changes requested):
I think it is worth considering refactoring `vortex-scalar` as well to truly support a "canonicalized vectors are our IR" world, but I'm not sure if it fits into this timeline?
